core/src/main/java/jenkins/console/ConsoleUrlProvider.java

```diff
@@ -0,0 +1,130 @@
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package jenkins.console;
+
+import edu.umd.cs.findbugs.annotations.CheckForNull;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import hudson.Extension;
+import hudson.Functions;
+import hudson.model.Describable;
+import hudson.model.Descriptor;
+import hudson.model.Run;
+import hudson.model.User;
+import java.util.List;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import jenkins.model.Jenkins;
+import org.jenkinsci.Symbol;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+import org.kohsuke.stapler.Stapler;
+
+/**
+ * Extension point that allows implementations to redirect build console links to a specified URL.
+ * <p>In order to produce links to console URLs in Jelly templates, use {@link Functions#getConsoleUrl}.
+ * @see Functions#getConsoleUrl
+ * @since TODO
+ */
+public interface ConsoleUrlProvider extends Describable<ConsoleUrlProvider> {
+    @Restricted(NoExternalUse.class)
+    Logger LOGGER = Logger.getLogger(ConsoleUrlProvider.class.getName());
+
+    /**
+     * Get a URL relative to the context path of Jenkins which should be used to link to the console for the specified build.
+     * <p>Should only be used in the context of serving an HTTP request.
+     * @param run the build
+     * @return the URL for the console for the specified build, relative to the context of Jenkins, or {@code null}
+     * if this implementation does not want to server a special console view for this build.
+     */
+    @CheckForNull String getConsoleUrl(Run<?, ?> run);
+
+    @Override
+    default Descriptor<ConsoleUrlProvider> getDescriptor() {
+        return Jenkins.get().getDescriptorOrDie(getClass());
+    }
+
+    /**
+     * Get a URL relative to the web server root which should be used to link to the console for the specified build.
+     * <p>Should only be used in the context of serving an HTTP request.
+     * <p>Use {@link Functions#getConsoleUrl} to obtain this link in a Jelly template.
+     * @param run the build
+     * @return the URL for the console for the specified build, relative to the web server root
+     */
+    static @NonNull String getRedirectUrl(Run<?, ?> run) {
+        ConsoleUrlProviderGlobalConfiguration globalConfig = ConsoleUrlProviderGlobalConfiguration.get();
+        List<ConsoleUrlProvider> providers = globalConfig.getProviders();
+        User currentUser = User.current();
+        if (currentUser != null) { // TODO: Should we add a configuration option to allow admins to suppress this behavior?
```

Comment:
In other words, do admins want to be able to restrict users from being able to configure things, or do they just want to set defaults? For now admins could use extension-filter on `ConsoleUrlProviderUserProperty.DescriptorImpl` if they really want full control, so I am not inclined to add additional options.

Reply:
Given no one seems to think this is necessary, I dropped the comment for now.